x/mongo/driver/topology/pool.go

```diff
@@ -275,7 +275,16 @@ func newPool(config poolConfig, connOpts ...ConnectionOption) *pool {
 
 // stale checks if a given connection's generation is below the generation of the pool
 func (p *pool) stale(conn *connection) bool {
-	return conn == nil || p.generation.stale(conn.desc.ServiceID, conn.generation)
+	if conn == nil {
+		return true
+	}
+	if atomic.LoadInt64(&p.generation.state) == generationDisconnected {
+		return true
+	}
```

Comment:
The `p.generation` is a substitute for the pool state because `p.generation.disconnect()` is called in `*pool.close()`, while locking `stateMu` may cause deadlock.